pkg/kds/store/sync.go

```diff
@@ -0,0 +1,139 @@
+package store
+
+import (
+	"context"
+
+	"github.com/Kong/kuma/pkg/core/resources/model"
+	"github.com/Kong/kuma/pkg/core/resources/registry"
+	"github.com/Kong/kuma/pkg/core/resources/store"
+)
+
+// SyncResourceStore extends ResourceStore with Sync method
+type SyncResourceStore interface {
+	store.ResourceStore
+	// Sync method takes 'upstream' as a basis and synchronize underlying store.
+	// It deletes all resources that absent in 'upstream', creates new resources that
+	// are not represented in store yet and updates the rest.
+	Sync(upstream model.ResourceList, fs ...SyncOptionFunc) error
+}
+
+type SyncOption struct {
+	Predicate func(r model.Resource) bool
+}
+
+type SyncOptionFunc func(*SyncOption)
+
+func NewSyncOptions(fs ...SyncOptionFunc) *SyncOption {
+	opts := &SyncOption{}
+	for _, f := range fs {
+		f(opts)
+	}
+	return opts
+}
+
+func PrefilterBy(predicate func(r model.Resource) bool) SyncOptionFunc {
+	return func(opts *SyncOption) {
+		opts.Predicate = predicate
+	}
+}
+
+type syncResourceStore struct {
+	store.ResourceStore
+}
+
+func NewSyncResourceStore(resourceStore store.ResourceStore) SyncResourceStore {
+	return &syncResourceStore{
+		ResourceStore: resourceStore,
+	}
+}
+
+func (s *syncResourceStore) Sync(upstream model.ResourceList, fs ...SyncOptionFunc) error {
```

Comment:
I think we should add logs to this method.
```
log.Info("Deleting a resource since it's no longer available in the upstream", "resourceKey", resourceKey)
```
etc.